SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -1,271 +1,215 @@
-			continue;
-		}
+	const WorkerAttributeSet OwningServerWorkerAttributeSet = { WriteWorkerId };
 
-		check(VirtualWorkerTranslator != nullptr);
-		const PhysicalWorkerName* DestinationWorkerId =
-			VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
-		if (DestinationWorkerId == nullptr)
+	for (const Worker_ComponentId& ComponentId : ComponentIds)
+	{
+		if (ComponentId == SpatialConstants::HEARTBEAT_COMPONENT_ID
+			|| ComponentId == SpatialConstants::GetClientAuthorityComponent(GetDefault<USpatialGDKSettings>()->UseRPCRingBuffer()))
 		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Error,
-				   TEXT("This worker is not assigned a virtual worker. This shouldn't happen! Worker: %s"), *WorkerId);
+			Acl.ComponentWriteAcl.Add(ComponentId, { { PossessingClientId } });
 			continue;
 		}
 
-		if (!StaticComponentView->HasAuthority(EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
+		if (ComponentId == SpatialConstants::ENTITY_ACL_COMPONENT_ID)
 		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Log,
-				   TEXT("Failed to update the EntityACL to match the authority intent; this worker lost authority over the EntityACL since "
-						"the request was queued."
-						" Source worker ID: %s. Entity ID %lld. Desination worker ID: %s."),
-				   *WorkerId, EntityId, **DestinationWorkerId);
-			CompletedRequests.Add(EntityId);
+			Acl.ComponentWriteAcl.Add(ComponentId, { SpatialConstants::UnrealServerAttributeSet });
 			continue;
 		}
 
-		TArray<Worker_ComponentId> ComponentIds;
-
-		EntityAcl* Acl = StaticComponentView->GetComponentData<EntityAcl>(EntityId);
-		Acl->ComponentWriteAcl.GetKeys(ComponentIds);
-
-		// Ensure that every component ID in ComponentPresence is set in the write ACL.
-		for (const auto& RequiredComponentId : ComponentPresenceComponent->ComponentList)
-		{
-			ComponentIds.AddUnique(RequiredComponentId);
-		}
-
-		// Get the client worker ID net-owning this Actor from the NetOwningClientWorker.
-		PhysicalWorkerName PossessingClientId =
-			NetOwningClientWorkerComponent->WorkerId.IsSet() ? NetOwningClientWorkerComponent->WorkerId.GetValue() : FString();
-
-		PendingRequests.Push(
-			AclWriteAuthorityRequest{ EntityId, *DestinationWorkerId, Acl->ReadAcl, { { PossessingClientId } }, ComponentIds });
-
-		CompletedRequests.Add(EntityId);
+		Acl.ComponentWriteAcl.Add(ComponentId, { OwningServerWorkerAttributeSet });
 	}
 
-	AclWriteAuthAssignmentRequests.RemoveAll([CompletedRequests](const Worker_EntityId& EntityId) {
-		return CompletedRequests.Contains(EntityId);
-	});
-
-	return PendingRequests;
+	const FWorkerComponentUpdate Update = Acl.CreateEntityAclUpdate();
+	return EntityComponentUpdate{ EntityId, ComponentUpdate(OwningComponentUpdatePtr(Update.schema_type), Update.component_id) };
 }
 
-void SpatialLoadBalanceEnforcer::QueueAclAssignmentRequest(const Worker_EntityId EntityId)
+void SpatialLoadBalanceEnforcer::PopulateDataStore(const Worker_EntityId EntityId)
```

Comment:
the LB enforcer maintains all its own state- complete decoupling from the rest of the world. This is part of the efforts to remove the static component view, and should make the LB enforcer far easier to understand when it does weird things

Reply:
It adds a bit of bloat to the enforcer now, but the state maintenance can be decoupled into its own class. Didn't do that now because AFAIK we're going to replace this all with strategy worker stuff.